src/commands/force/mdapi/beta/retrieve.ts

```diff
@@ -0,0 +1,304 @@
+    unzip: flags.boolean({
+      char: 'z',
+      description: messages.getMessage('flags.unzip'),
+      longDescription: messages.getMessage('flagsLong.unzip'),
+    }),
+    wait: flags.minutes({
+      char: 'w',
+      description: messages.getMessage('flags.wait'),
+      longDescription: messages.getMessage('flagsLong.wait'),
+      default: Duration.minutes(1440), // 24 hours is a reasonable default versus -1 (no timeout)
+    }),
+    apiversion: flags.builtin({
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore force char override for backward compat
+      char: 'a',
+      description: messages.getMessage('flags.apiversion'),
+      longDescription: messages.getMessage('flagsLong.apiversion'),
+    }),
+    verbose: flags.builtin({
+      description: messages.getMessage('flags.verbose'),
+      longDescription: messages.getMessage('flagsLong.verbose'),
+    }),
+  };
+
+  protected retrieveResult: RetrieveResult;
+  private sourceDir: string;
+  private retrieveTargetDir: string;
+  private zipFileName: string;
+  private unzip: boolean;
+  private wait: Duration;
+  private isAsync: boolean;
+  private mdapiRetrieve: MetadataApiRetrieve;
+
+  public async run(): Promise<RetrieveCommandResult | RetrieveCommandAsyncResult> {
+    await this.retrieve();
+    this.resolveSuccess();
+    return this.formatResult();
+  }
+
+  protected async retrieve(): Promise<void> {
+    const packagenames = this.getFlag<string[]>('packagenames');
+    if (!packagenames) {
+      this.sourceDir = this.resolveRootDir(this.getFlag<string>('sourcedir'));
+    }
+    this.retrieveTargetDir = this.resolveOutputDir(this.getFlag<string>('retrievetargetdir'));
+    const manifest = this.resolveManifest(this.getFlag<string>('unpackaged'));
+    const singlePackage = this.getFlag<boolean>('singlepackage');
+    this.zipFileName = this.getFlag<string>('zipfilename') || 'unpackaged.zip';
+    this.unzip = this.getFlag<boolean>('unzip');
+    this.wait = this.getFlag<Duration>('wait');
+    this.isAsync = this.wait.quantity === 0;
+
+    if (singlePackage && packagenames?.length > 1) {
+      throw SfdxError.create('@salesforce/plugin-source', 'md.retrieve', 'InvalidPackageNames', [
+        packagenames.toString(),
+      ]);
+    }
+
+    this.ux.startSpinner(spinnerMessages.getMessage('retrieve.main', [this.org.getUsername()]));
+    this.ux.setSpinnerStatus(spinnerMessages.getMessage('retrieve.componentSetBuild'));
+
+    this.componentSet = await ComponentSetBuilder.build({
+      apiversion: this.getFlag<string>('apiversion'),
+      packagenames,
+      sourcepath: this.sourceDir ? [this.sourceDir] : undefined,
+      manifest: manifest && {
+        manifestPath: manifest,
+        directoryPaths: [],
+      },
+    });
+
+    await this.lifecycle.emit('preretrieve', { packageXmlPath: manifest });
+
+    this.ux.setSpinnerStatus(spinnerMessages.getMessage('retrieve.sendingRequest', [this.componentSet.apiVersion]));
```

Comment:
for consolidation, maybe this belongs on sourceCommand: 
https://github.com/salesforcecli/plugin-source/blob/5f0ad449a6880d761a233c3191e3759a6c0b72cc/src/deployCommand.ts#L212